MineSweeper/Assets/Scripts/Box.cs

```diff
@@ -0,0 +1,184 @@
+            (1024f / (scale+1)) / flag.rect.size.y, 1f);
+        o.GetComponent<Box>().state = 2;
+        
+    }
+                
+    void ChangeSprite(GameObject o)
+    {
+        Sprite image = number[o.GetComponent<Box>().neighbour];
+        o.GetComponent<SpriteRenderer>().sprite = image;
+        o.GetComponent<SpriteRenderer>().color = Color.white;
+        o.transform.localScale = new Vector3((1024f / (scale+1)) / image.rect.size.x,
+            (1024f / (scale+1)) / image.rect.size.y, 1f);
+        o.GetComponent<Box>().state = 2;
+        
+        opened += 1;
+        Debug.Log(o.name);
+        Debug.Log(opened);
+    }
+    
+    void ChangeSpriteReset(GameObject o)
+    {
+        Sprite image = boxColor;
+        o.GetComponent<SpriteRenderer>().sprite = image;
+        o.GetComponent<SpriteRenderer>().color = Color.white;
+        o.transform.localScale = new Vector3((1024f / (scale+1)) / image.rect.size.x,
+            (1024f / (scale+1)) / image.rect.size.y, 1f);
+        o.GetComponent<Box>().state = 0;
+    }
+
+    public void ClickOn()
+    {
+       
+        GameObject self = GameObject.Find(name);
+        if (state == 0)
+        {
+            if (Input.GetMouseButtonUp(1))
+            {
+                ChangeSpriteFlag(self);
+            }
+
+            if (Input.GetMouseButtonUp(0))
+            {
+                if (mine)
+                {
+                    ChangeSpriteMine(self);
+                    EndGameController.win = false;
+                    die = true;
+                }
+                else
+                {
+                    ChangeSprite(self);
+                    // first find the x and y
+                    (int x, int y) = findLocation(GameObject.Find(name));
+                    // bfs to expand neighbour
+                    List<(int, int)> queue = new List<(int, int)>();
+                    if (neighbour == 0)
+                    {
+                        queue.Add((x,y));
+                    }
+                    while (queue.Count > 0)
+                    {
+                        ( x,  y)= queue[0];
+                        queue.RemoveAt(0);
+                        // get the valid neighbours
+                        GameObject curBox;
+                        for (int i = -1; i < 2; i++)
+                        {
+                            if (x + i < 0 || x + i >= scale)
+                            {
+                                continue;
+                            }
+                            for (int j = -1; j < 2; j++)
+                            {
+                                if (y + j >= 0 && y + j < scale)
+                                {
+                                    if (i == 0 && j == 0)
+                                    {
+                                        continue;
+                                    }
+                                    curBox = GameObject.Find("box" + (x+i) + "_" + (y+j));
```

Comment:
its incredibly inefficient to search by name. please create an array of model. you need to rebuild this with MVC to achieve readability and performance.